crates/broker/src/order_picker.rs

```diff
@@ -547,42 +566,148 @@ where
             exec_limit_cycles,
             exec_limit_cycles / 1_000_000
         );
-        // TODO add a future timeout here to put a upper bound on how long to preflight for
-        let proof_res = match self
-            .prover
-            .preflight(
-                &image_id,
-                &input_id,
-                vec![],
-                /* TODO assumptions */ Some(exec_limit_cycles),
-                &order_id,
-            )
-            .await
-        {
-            Ok(res) => {
-                tracing::debug!(
-                    "Preflight execution of {order_id} with session id {} and {} mcycles completed in {} seconds",
-                    res.id,
-                    res.stats.total_cycles / 1_000_000,
-                    res.elapsed_time
-                );
-                res
+
+        // Create cache key based on input type
+        let image_id = Digest::from(order.request.requirements.imageId.0);
+        let cache_key = match order.request.input.inputType {
+            RequestInputType::Url => {
+                let input_url = std::str::from_utf8(&order.request.input.data)
+                    .context("input url is not utf8")
+                    .map_err(|e| OrderPickerErr::FetchInputErr(Arc::new(e)))?
+                    .to_string();
+                PreflightCacheKey { image_id, input: InputCacheKey::Url(input_url) }
+            }
+            RequestInputType::Inline => {
+                // For inline inputs, use SHA256 hash of the data
+                let mut hasher = Sha256::new();
+                Sha2Digest::update(&mut hasher, &order.request.input.data);
+                let input_hash: [u8; 32] = hasher.finalize().into();
+                PreflightCacheKey { image_id, input: InputCacheKey::Hash(input_hash) }
             }
-            Err(err) => match err {
-                ProverError::ProvingFailed(ref err_msg)
-                    if err_msg.contains("Session limit exceeded") =>
-                {
+            RequestInputType::__Invalid => {
+                return Err(OrderPickerErr::UnexpectedErr(Arc::new(anyhow::anyhow!(
+                    "Unknown input type: {:?}",
+                    order.request.input.inputType
+                ))));
+            }
+        };
+
+        // Loop while the cached result is skipped and has a lower exec limit than the current order.
+        let preflight_result = loop {
+            let prover = &self.prover;
+            let config = &self.config;
+            let request = &order.request;
+            let order_id_clone = &order_id;
+
+            let result = self
+                .preflight_cache
+                .try_get_with(cache_key.clone(), async move {
+                    // Upload image and input only if not cached
```

Comment:
Would be nice to add a note on the magic here that makes it so that even in-progress preflights result in a cache hit:

https://docs.rs/moka/latest/moka/sync/struct.Cache.html#concurrent-calls-on-the-same-key-2